lib/phoenix/channel.ex

```diff
@@ -204,29 +202,85 @@ defmodule Phoenix.Channel do
               {:shutdown, :left | :closed} |
               term
 
+
   defmacro __using__(_) do
     quote do
       @behaviour unquote(__MODULE__)
+      @on_definition unquote(__MODULE__)
+      @before_compile unquote(__MODULE__)
       import unquote(__MODULE__)
       import Phoenix.Socket, only: [assign: 3]
+      Module.register_attribute(__MODULE__, :phoenix_handle_outs, accumulate: true)
+
+      @phoenix_intercepts []
 
       def handle_in(_event, _message, socket) do
         {:noreply, socket}
       end
 
-      def handle_out(event, message, socket) do
-        push(socket, event, message)
+      def handle_info(_message, socket), do: {:noreply, socket}
+
+      def terminate(_reason, _socket), do: :ok
+
+      defoverridable handle_info: 2, handle_in: 3, terminate: 2
+    end
+  end
+
+  defmacro __before_compile__(env) do
+    intercepts = Module.get_attribute(env.module, :phoenix_intercepts)
+    handle_outs = Module.get_attribute(env.module, :phoenix_handle_outs)
```

Comment:
We don't need to store the handle outs, do we? We can emit the warning directly from `on_definition`, which would even allow us to use the file and line info.